packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -445,6 +454,273 @@ class CupertinoPopupSurface extends StatelessWidget {
+// corresponding method will be called on all avatars that nest it.
+abstract class _ActionSheetDragAvatar {
+  // A pointer has entered this region.
+  //
+  // This includes:
+  //
+  //  * The pointer has moved into this region from outside.
+  //  * The point has contacted the screen in this region. In this case, this
+  //    method is called as soon as the pointer down event occurs regardless of
+  //    whether the gesture wins the arena immediately.
+  void didEnter();
+
+  // A pointer has exited this region.
+  //
+  // This includes:
+  //  * The pointer has moved out of this region.
+  //  * The pointer is no longer in contact with the screen.
+  //  * The pointer is canceled.
+  //  * The gesture loses the arena.
+  //  * The gesture is completed. In this case, this method is called immediately
+  //    before [didConfirm].
+  void didLeave();
+
+  // The drag gesture is completed in this region.
+  //
+  // This method is called immediately after a [didLeave].
+  void didConfirm();
+}
+
+// Recognizes sliding taps and thereupon interacts with
+// `_ActionSheetDragAvatar`s.
+class _AvatarSelectionGestureRecognizer extends GestureRecognizer {
+  _AvatarSelectionGestureRecognizer({super.debugOwner, required this.hitTest})
+    : _slidingTap = _SlidingTapGestureRecognizer(debugOwner: debugOwner) {
+    _slidingTap
+      ..onDown = _onDown
+      ..onResponsiveUpdate = _onUpdate
+      ..onResponsiveEnd = _onEnd
+      ..onCancel = _onCancel;
+  }
+
+  final _HitTester hitTest;
+
+  final List<_ActionSheetDragAvatar> _currentAvatars = <_ActionSheetDragAvatar>[];
+  final _SlidingTapGestureRecognizer _slidingTap;
+
+  @override
+  void acceptGesture(int pointer) {
+    _slidingTap.acceptGesture(pointer);
+  }
+
+  @override
+  void rejectGesture(int pointer) {
+    _slidingTap.rejectGesture(pointer);
+  }
+
+  @override
+  void addPointer(PointerDownEvent event) {
+    _slidingTap.addPointer(event);
+  }
+
+  @override
+  void addPointerPanZoom(PointerPanZoomStartEvent event) {
+    _slidingTap.addPointerPanZoom(event);
+  }
+
+  @override
+  void dispose() {
+    _slidingTap.dispose();
+    super.dispose();
+  }
+
+  // Collect the `_ActionSheetDragAvatar`s that are currently hit by the
+  // pointer, check whether the current avatars have changed, and invoke their
+  // methods if necessary.
+  void _updateDrag(Offset pointerPosition) {
+    final HitTestResult result = hitTest(pointerPosition);
+
+    final List<_ActionSheetDragAvatar> foundAvatars = <_ActionSheetDragAvatar>[];
+    for (final HitTestEntry entry in result.path) {
+      if (entry.target case final RenderMetaData target) {
+        if (target.metaData is _ActionSheetDragAvatar) {
+          foundAvatars.add(target.metaData as _ActionSheetDragAvatar);
+        }
+      }
+    }
+
+    // Compare whether the active avatar has changed by simply comparing
+    // the first (inner-most) avatar of the nest.
```

Comment:
are we assumed that _currentAvatars does not intersect foundAvatars? if so, maybe add some more explanation

Reply:
I've explained it in the doc of `_ActionSheetDragAvatar`. 

```
// Multiple `_ActionSheetDragAvatar`s might be nested.
// `_AvatarSelectionGestureRecognizer` uses a simple algorithm that only
// compares if the inner-most avatar has changed (which suffices our use case).
// Semantically, this means that all outer avatars will be treated as identical
// to the inner-most one, i.e. when the pointer enters or leaves an avatar, the
// corresponding method will be called on all avatars that nest it.
```

I can add reference to the doc here to make it clearer.